File: server-windows/src/Protocol/WireProtocol.cs
using System;
using System.IO;
using System.Text;

namespace OpenWirelessDisplay.Protocol;

/// <summary>
/// Protocolo binario compartido servidor (Windows) <-> cliente (Android).
/// Cada mensaje: [1 byte Type][4 bytes BE Length][Length bytes Payload].
/// Version del protocolo: 1.
/// </summary>
public static class WireProtocol
{
    public const byte ProtocolVersion = 1;

    // Nombre de servicio mDNS/DNS-SD anunciado por el servidor.
    public const string ServiceType = "_openwdisplay._tcp";
    public const string ServiceDomain = "local.";

    // Puerto TCP por defecto (control + stream multiplexados en la misma conexion).
    public const int DefaultPort = 7345;

    // ---- Cliente -> Servidor ----
    public const byte MsgHello = 0x01;   // payload: JSON {clientName, protocol}
    public const byte MsgPin = 0x02;     // payload: UTF8 PIN (6 digitos)
    public const byte MsgInput = 0x10;   // payload: [1 byte action][float x][float y] (BE)
    public const byte MsgBye = 0x7F;     // sin payload

    // ---- Servidor -> Cliente ----
    public const byte MsgHelloAck = 0x81; // payload: JSON {needPin, serverName, protocol}
    public const byte MsgPinOk = 0x82;    // payload: JSON {width, height, fps}
    public const byte MsgPinFail = 0x83;  // payload: JSON {reason, attemptsLeft}
    public const byte MsgFrame = 0x90;    // payload: [4 BE w][4 BE h][JPEG bytes]
    public const byte MsgError = 0xFE;    // payload: UTF8 mensaje

    // Acciones de input (touch/mouse).
    public const byte InputMove = 0x00;
    public const byte InputDown = 0x01;
    public const byte InputUp = 0x02;

    public static void WriteMessage(Stream stream, byte type, byte[]? payload)
    {
        payload ??= Array.Empty<byte>();
        Span<byte> header = stackalloc byte[5];
        header[0] = type;
        WriteInt32BE(header.Slice(1), payload.Length);
        stream.Write(header);
        if (payload.Length > 0)
            stream.Write(payload, 0, payload.Length);
        stream.Flush();
    }

    public static void WriteMessage(Stream stream, byte type, string utf8) =>
        WriteMessage(stream, type, Encoding.UTF8.GetBytes(utf8));

    /// <summary>Lee un mensaje completo. Devuelve false si la conexion se cerro.</summary>
    public static bool ReadMessage(Stream stream, out byte type, out byte[] payload)
    {
        type = 0;
        payload = Array.Empty<byte>();
        var header = new byte[5];
        if (!ReadExact(stream, header, 0, 5)) return false;
        type = header[0];
        int len = ReadInt32BE(header, 1);
        if (len < 0 || len > 64 * 1024 * 1024) throw new InvalidDataException($"Longitud invalida: {len}");
        payload = new byte[len];
        if (len > 0 && !ReadExact(stream, payload, 0, len)) return false;
        return true;
    }

    private static bool ReadExact(Stream stream, byte[] buffer, int offset, int count)
    {
        int read = 0;
        while (read < count)
        {
            int n = stream.Read(buffer, offset + read, count - read);
            if (n <= 0) return false;
            read += n;
        }
        return true;
    }

    public static void WriteInt32BE(Span<byte> dst, int value)
    {
        dst[0] = (byte)(value >> 24);
        dst[1] = (byte)(value >> 16);
        dst[2] = (byte)(value >> 8);
        dst[3] = (byte)value;
    }

    public static int ReadInt32BE(byte[] src, int offset) =>
        (src[offset] << 24) | (src[offset + 1] << 16) | (src[offset + 2] << 8) | src[offset + 3];

    public static float ReadFloatBE(byte[] src, int offset)
    {
        Span<byte> tmp = stackalloc byte[4];
        tmp[0] = src[offset + 3];
        tmp[1] = src[offset + 2];
        tmp[2] = src[offset + 1];
        tmp[3] = src[offset];
        return BitConverter.ToSingle(tmp);
    }
}

File: client-android/app/src/main/java/com/openwdisplay/client/WireProtocol.kt
package com.openwdisplay.client

import java.io.DataInputStream
import java.io.InputStream
import java.io.OutputStream
import java.nio.ByteBuffer
import java.nio.ByteOrder

/**
 * Protocolo binario compartido con el servidor Windows (.NET).
 * Mensaje: [1 byte Type][4 bytes BE Length][Length bytes Payload].
 * Debe mantenerse en paridad con server-windows/src/Protocol/WireProtocol.cs.
 */
object WireProtocol {
    const val PROTOCOL_VERSION = 1
    const val SERVICE_TYPE = "_openwdisplay._tcp."
    const val DEFAULT_PORT = 7345

    // Cliente -> Servidor
    const val MSG_HELLO: Int = 0x01
    const val MSG_PIN: Int = 0x02
    const val MSG_INPUT: Int = 0x10
    const val MSG_BYE: Int = 0x7F

    // Servidor -> Cliente
    const val MSG_HELLO_ACK: Int = 0x81
    const val MSG_PIN_OK: Int = 0x82
    const val MSG_PIN_FAIL: Int = 0x83
    const val MSG_FRAME: Int = 0x90
    const val MSG_ERROR: Int = 0xFE

    // Acciones de input
    const val INPUT_MOVE: Byte = 0x00
    const val INPUT_DOWN: Byte = 0x01
    const val INPUT_UP: Byte = 0x02

    data class Message(val type: Int, val payload: ByteArray)

    fun writeMessage(out: OutputStream, type: Int, payload: ByteArray = ByteArray(0)) {
        val header = ByteArray(5)
        header[0] = type.toByte()
        header[1] = (payload.size ushr 24).toByte()
        header[2] = (payload.size ushr 16).toByte()
        header[3] = (payload.size ushr 8).toByte()
        header[4] = payload.size.toByte()
        out.write(header)
        if (payload.isNotEmpty()) out.write(payload)
        out.flush()
    }

    fun writeMessage(out: OutputStream, type: Int, utf8: String) =
        writeMessage(out, type, utf8.toByteArray(Charsets.UTF_8))

    /** Lee un mensaje completo o lanza EOFException si la conexion se cierra. */
    fun readMessage(input: DataInputStream): Message {
        val type = input.readUnsignedByte()
        val len = input.readInt() // big-endian
        require(len in 0..(64 * 1024 * 1024)) { "Longitud invalida: $len" }
        val payload = ByteArray(len)
        input.readFully(payload)
        return Message(type, payload)
    }

    fun encodeInput(action: Byte, normX: Float, normY: Float): ByteArray {
        val buf = ByteBuffer.allocate(9).order(ByteOrder.BIG_ENDIAN)
        buf.put(action)
        buf.putFloat(normX)
        buf.putFloat(normY)
        return buf.array()
    }

    fun readInt32BE(src: ByteArray, offset: Int): Int =
        ((src[offset].toInt() and 0xFF) shl 24) or
        ((src[offset + 1].toInt() and 0xFF) shl 16) or
        ((src[offset + 2].toInt() and 0xFF) shl 8) or
        (src[offset + 3].toInt() and 0xFF)
}
